### PR TITLE
Add refactor notify

### PR DESCRIPTION
**Pull request by jmarolf**
dependent upon https://github.com/dotnet/project-system/pull/4983

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Refactor/VisualStudioRefactorNotifyService.Instance.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+                }
+
+                if (!(projectHierarchy is IVsHierarchyRefactorNotify refactorNotify))
+                {
+                    return false;
+                }
+
+                var ids = GetIdsForFiles(projectHierarchy, filePaths).ToArray();
+
+                // call OnBeforeGlobalSymbolRenamed
```
Comment (davidwengier) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Refactor/VisualStudioRefactorNotifyService.Instance.cs]: Nit: Comment is wrong

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Refactor/VisualStudioRefactorNotifyService.Instance.cs]: I'd vote for removing this comment and the one above from which it was presumably copied. They're stating the obvious.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Refactor/VisualStudioRefactorNotifyService.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        public async Task<bool> TryOnBeforeGlobalSymbolRenamedAsync(string projectPath, IEnumerable<string> filePaths, string rqName, string newName)
+        {
+            if (projectPath is null)
+                throw new System.ArgumentNullException(nameof(projectPath));
+            if (filePaths is null)
+                throw new System.ArgumentNullException(nameof(filePaths));
+            if (rqName is null)
+                throw new System.ArgumentNullException(nameof(rqName));
+            if (newName is null)
+                throw new System.ArgumentNullException(nameof(newName));
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Refactor/VisualStudioRefactorNotifyService.cs]: These four should use `Requires.NotNull`.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Refactor/VisualStudioRefactorNotifyService.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        public async Task<bool> TryOnAfterGlobalSymbolRenamedAsync(string projectPath, IEnumerable<string> filePaths, string rqName, string newName)
+        {
+            if (projectPath is null)
+                throw new System.ArgumentNullException(nameof(projectPath));
+            if (filePaths is null)
+                throw new System.ArgumentNullException(nameof(filePaths));
+            if (rqName is null)
+                throw new System.ArgumentNullException(nameof(rqName));
+            if (newName is null)
+                throw new System.ArgumentNullException(nameof(newName));
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Refactor/VisualStudioRefactorNotifyService.cs]: Ditto.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Refactor/IRefactorNotifyService.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Refactor
+{
+    internal interface IRefactorNotifyService
+    {
+        Task<bool> TryOnBeforeGlobalSymbolRenamedAsync(string projectPath, IEnumerable<string> filePaths, string rqName, string newName);
+        Task<bool> TryOnAfterGlobalSymbolRenamedAsync(string projectPath, IEnumerable<string> filePaths, string rqName, string newName);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Refactor/IRefactorNotifyService.cs]: These return values are always ignored.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Refactor/IRefactorNotifyService.cs [hunk 5]
```diff
@@ -2,13 +2,12 @@
...
-using Microsoft.CodeAnalysis;
 
 namespace Microsoft.VisualStudio.ProjectSystem.Refactor
 {
     internal interface IRefactorNotifyService
     {
-        Task<bool> TryOnBeforeGlobalSymbolRenamedAsync(string projectPath, IEnumerable<string> filePaths, string rqName, string newName);
-        Task<bool> TryOnAfterGlobalSymbolRenamedAsync(string projectPath, IEnumerable<string> filePaths, string rqName, string newName);
+        Task TryOnBeforeGlobalSymbolRenamedAsync(string projectPath, IEnumerable<string> filePaths, string rqName, string newName);
+        Task TryOnAfterGlobalSymbolRenamedAsync(string projectPath, IEnumerable<string> filePaths, string rqName, string newName);
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Refactor/IRefactorNotifyService.cs]: Should these still be `Try*`?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Refactor/IRefactorNotifyService.cs (hunk 4 above)
Comment (jmarolf) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Refactor/IRefactorNotifyService.cs]: Yep I thought we might want to handle the failure case when I was writing this but I realized later it didn't matter

**Comment by jmarolf** [comment]
> I would have thought Roslyn would trigger IRefactorNotify when you tell Roslyn to rename the type ¯\_(ツ)_/¯

IRefactorNotify is a visual studio concept and roslyn is a platform-agnistic api (at least at the layer that roslyn lives at). Using roslyns code model apis will call refactor notify but it is not cancelable

**Pull request merged**